core_lib/src/interface/scribblearea.cpp

```diff
@@ -274,60 +243,114 @@ void ScribbleArea::removeOnionSkinsCacheAround(int frameNumber)
             onionFrameNumber = layer->getPreviousFrameNumber(onionFrameNumber, isOnionAbsolute);
             if (onionFrameNumber < 0) break;
 
-            auto cacheKeyIter = mPixmapCacheKeys.find(static_cast<unsigned int>(onionFrameNumber));
-            if (cacheKeyIter != mPixmapCacheKeys.end())
-            {
-                QPixmapCache::remove(cacheKeyIter.value());
-                unsigned int key = cacheKeyIter.key();
-                mPixmapCacheKeys.remove(key);
-            }
+            invalidateCacheForFrame(onionFrameNumber);
         }
     }
 }
 
-void ScribbleArea::updateAllFrames()
+void ScribbleArea::invalidateAllCache()
 {
     QPixmapCache::clear();
     mPixmapCacheKeys.clear();
-    setAllDirty();
+    invalidateLayerPixmapCache();
 
     update();
 }
 
-void ScribbleArea::updateAllVectorLayersAtCurrentFrame()
+void ScribbleArea::invalidateCacheForFrame(int frameNumber)
+{
+    auto cacheKeyIter = mPixmapCacheKeys.find(static_cast<unsigned int>(frameNumber));
+    if (cacheKeyIter != mPixmapCacheKeys.end())
+    {
+        QPixmapCache::remove(cacheKeyIter.value());
+        unsigned int key = cacheKeyIter.key();
+        mPixmapCacheKeys.remove(key);
+    }
+}
+
+void ScribbleArea::onPlayStateChanged()
 {
-    updateAllVectorLayersAt(mEditor->currentFrame());
+    bool stoppedPlaying = !mEditor->playback()->isPlaying();
+    int currentFrame = mEditor->currentFrame();
+    if (stoppedPlaying &&
+       (mPrefs->isOn(SETTING::PREV_ONION) ||
+        mPrefs->isOn(SETTING::NEXT_ONION))) {
+        invalidateOnionSkinsCacheAround(currentFrame);
+        invalidateCacheForFrame(currentFrame);
+    }
+
+    updateFrame(currentFrame);
 }
 
-void ScribbleArea::updateAllVectorLayersAt(int frameNumber)
+void ScribbleArea::onScrubbed(int frameNumber)
 {
-    for (int i = 0; i < mEditor->object()->getLayerCount(); i++)
-    {
-        Layer* layer = mEditor->object()->getLayer(i);
-        if (layer->type() == Layer::VECTOR)
-        {
-            VectorImage* vectorImage = currentVectorImage(layer);
-            if (vectorImage != nullptr)
-            {
-                vectorImage->modification();
-            }
-        }
+    invalidateCacheForDirtyFrames();
+    if (mPrefs->isOn(SETTING::PREV_ONION) || mPrefs->isOn(SETTING::NEXT_ONION)) {
+        invalidateOnionSkinsCacheAround(frameNumber);
     }
     updateFrame(frameNumber);
 }
 
-void ScribbleArea::setModified(int layerNumber, int frameNumber)
+void ScribbleArea::onCurrentFrameModified()
 {
-    Layer* layer = mEditor->object()->getLayer(layerNumber);
-    if (layer)
+    onFrameModified(mEditor->currentFrame());
+}
+
+void ScribbleArea::onFrameModified(int frameNumber)
+{
+    mCurrentCacheInvalid = true;
```

Comment:
Question. Why raising a flag here rather than clearing the cache directly?

Reply:
Hmm good question, I guess it's a remnant from before where I had multiple places where mCurrentCacheInvalid was set to true but yeah it doesn't make sense anymore.